server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java

```diff
@@ -49,21 +50,56 @@ protected boolean searchWithCollector(
         boolean hasFilterCollector,
         boolean hasTimeout
     ) throws IOException {
-        return searchWithCollectorManager(searchContext, searcher, query, collectors, hasFilterCollector, hasTimeout);
+        // create the top docs collector
+        final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(searchContext, hasFilterCollector);
+        return searchWithCollector(
+            searchContext,
+            searcher,
+            query,
+            collectors,
+            topDocsFactory,
+            hasFilterCollector,
+            hasTimeout,
+            topDocsFactory.shouldRescore()
+        );
+    }
+
+    protected boolean searchWithCollector(
+        SearchContext searchContext,
+        ContextIndexSearcher searcher,
+        Query query,
+        LinkedList<QueryCollectorContext> collectors,
+        QueryCollectorContext queryCollectorContext,
+        boolean hasFilterCollector,
+        boolean hasTimeout,
+        boolean shouldRescore
+    ) throws IOException {
+        return searchWithCollectorManager(
+            searchContext,
+            searcher,
+            query,
+            collectors,
+            queryCollectorContext,
+            hasFilterCollector,
+            hasTimeout,
+            shouldRescore
+        );
     }
 
     private static boolean searchWithCollectorManager(
         SearchContext searchContext,
         ContextIndexSearcher searcher,
         Query query,
         LinkedList<QueryCollectorContext> collectorContexts,
+        QueryCollectorContext queryCollectorContext,
         boolean hasFilterCollector,
-        boolean timeoutSet
+        boolean timeoutSet,
+        boolean shouldRescore
     ) throws IOException {
-        // create the top docs collector last when the other collectors are known
-        final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(searchContext, hasFilterCollector);
-        // add the top docs collector, the first collector context in the chain
-        collectorContexts.addFirst(topDocsFactory);
+        if (Objects.nonNull(queryCollectorContext)) {
```

Comment:
Should we use `EMPTY_CONTEXT` instead of dealing with `null`s?

```
    public static final QueryCollectorContext EMPTY_CONTEXT = new QueryCollectorContext("empty") {
        @Override
        Collector create(Collector in) throws IOException {
            return EMPTY_COLLECTOR;
        }

        @Override
        CollectorManager<?, ReduceableSearchResult> createManager(CollectorManager<?, ReduceableSearchResult> in) throws IOException {
            return new CollectorManager<Collector, ReduceableSearchResult>() {
                @Override
                public Collector newCollector() throws IOException {
                    return EMPTY_COLLECTOR;
                }

                @Override
                public ReduceableSearchResult reduce(Collection<Collector> collectors) throws IOException {
                    return result -> {};
                }
            };
        }
    };
```

Reply:
let me try this approach, it would be nice if that works, I don't like dealing with `null` either

Reply:
Added EMPTY_CONTEXT static instance